Fix broken tests

PeterSurda opened the pull request:
- setting `BITMESSAGE_HOME` doesn't always work, so we wrote a function to do this
- `logger` and `BMConfigParser` now clear old config when loading new
- add method for flushing log files
- rework logic in `test_logger`
- detect thread names on UNIX-like OSes
- be more accurate when starting or stopping a process, better waiting mechanisms
- add extra cleanup during `TestProcess` start

g1itch commented:
Ooogh, it looks sick ):

g1itch commented on .travis.yml:
```diff
@@ -1,6 +1,8 @@
 python:
   - "2.7"
+env:
+  - BITMESSAGE_HOME=/tmp
```
Sad. I didn't planned to bound the tests to travis.

PeterSurda replied:
I agree but I don't think it can be fixed without a lot of refactoring.

g1itch commented on src/helper_addressbook.py:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python
```
Why shebang here?

g1itch replied:
Another helper ):

PeterSurda replied:
Ok.

PeterSurda replied:
I mean, I agree about the shebang, but the addressbook code shouldn't be in the UI like it is now. What other options are there? Not another thread to deal with addressbook?

g1itch replied:
I agree that it's cool to have all CRUD tasks (on messages, addressbook and black/white lists) in the main package to use them in `api`, `bitmessageqt` and any other frontend. Also I was working on the single helper with all SQL queries in #1534 because it is bad to have SQL scattered amond the code. It needs heavy testing. Currently it breaks `bitmessageqt` because I missed some type cast or check.

But hey, you need this helper now only because of `BITMESSAGE_HOME` in env and the absence of cleanup.

Another thread is the last thing I would suggest. But recently I thought about using multiprocessing for test cases which use potentially dangerous imports.

g1itch commented on src/tests/test_process.py:
```diff
@@ -110,7 +172,16 @@ def test_shutdown(self):
-        cls._cleanup_files()
+        try:
+            if cls.process.is_running():
+                cls.process.kill()
```
This looks like a cheat. `TestProcessProto` does the same: https://github.com/Bitmessage/PyBitmessage/blob/v0.6/src/tests/test_process.py#L73

PeterSurda replied:
1. it isn't the same, because `TestProcessProto` first calls `_stop_process` to send a `SIGTERM`. This expects the program to already have closed on its own, and just send `SIGKILL` in an emergency.
2. it allows you to put custom handlers into `TimeoutExpired` and `NoSuchProcess` if you want the tests to be more strict. `TimeoutExpired` means that there is something very wrong as it's blocking a `SIGKILL`.
3. However, the main point here is line 178 as previously, it wasn't waiting for the program to finish after `SIGKILL`, which I think lead unpredictably to problems.

g1itch replied:
OK, I get it:
- it doesn't disable the test because it's failing by it's own if process not stops.
- it is not the same as super

g1itch commented:
I don't think that increasing timeouts gonna help.

g1itch commented on src/bitmessageqt/tests/support.py:
```diff
@@ -12,6 +13,10 @@ class TestSupport(TestBase):
 
+    def tearDown(self):
+        """Delete the support address from addressbook"""
+        delete_from_addressbook(self.SUPPORT_ADDRESS)
```
core tests run from appdata

PeterSurda replied:
Maybe yes, but noone is going to do that.

g1itch replied:
I did it in one of my rebased branches. It did not help, because cleanup is the least part of the problem.

g1itch replied:
Did you try this?
```patch
--- a/src/tests/test_process.py
+++ b/src/tests/test_process.py
@@ -33,6 +33,7 @@ class TestProcessProto(unittest.TestCase):
     def setUpClass(cls):
         """Setup environment and start pybitmessage"""
         cls.home = os.environ['BITMESSAGE_HOME'] = tempfile.gettempdir()
+        cls._cleanup_files()
         put_signal_file(cls.home, 'unittest.lock')
         subprocess.call(cls._process_cmd)  # nosec
         time.sleep(5)

```

PeterSurda replied:
ok let me try

g1itch commented on src/bitmessageqt/tests/support.py:
```diff
@@ -11,6 +13,17 @@ class TestSupport(TestBase):
     """A test case for support module"""
     SUPPORT_ADDRESS = 'BM-2cUdgkDDAahwPAU6oD2A7DnjqZz3hgY832'
     SUPPORT_SUBJECT = 'Support request'
+    SUPPORT_MY_LABEL = 'My new address'
```
_translate ?

PeterSurda replied:
ok

g1itch commented on src/tests/test_process.py:
```diff
@@ -94,8 +129,35 @@ def _test_threads(self):
-            len(self.process.threads()), self._threads_count)
+        #   - Uploader
+
+        self.longMessage = True
```
I cannot find where it's used

g1itch replied:
Oh, it's from unittest.TestCase. Got it.

PeterSurda replied:
Yes it's so that we have more info about the failures.

g1itch commented on src/tests/test_process.py:
```diff
@@ -94,8 +129,35 @@ def _test_threads(self):
+        #   - Uploader
+
+        self.longMessage = True
+
```
If this hunk is platform dependent, maybe here should be some condition?

PeterSurda replied:
I haven't tested it on Windows, but I assumed the result will be empty. Maybe it can be wrapped in try/except. The output is only used if there is an assertion error, it's not an assertion itself.

g1itch replied:
https://travis-ci.org/github/g1itch/PyBitmessage/jobs/738510444

g1itch reviewed (approved):
At your discretion

g1itch commented:
See more: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738338172

PeterSurda commented:
> I don't think that increasing timeouts gonna help.

Noone is working on a proper fix and other work can't be done until the tests succeed.

PeterSurda commented:
> See more: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738338172
> I'm waiting.

`test_config_defaults` isn't waiting for the process to finish.

g1itch commented:
> > I don't think that increasing timeouts gonna help.
> 
> Noone is working on a proper fix and other work can't be done until the tests succeed.

I don't remember any shutdown tests failing in your test environment or current xenial travis setup. APIShutdown failed on windows and on my recent fedora31 container.

g1itch commented on src/tests/test_process.py:
```diff
@@ -34,8 +53,22 @@ def setUpClass(cls):
+                pass
+            time.sleep(1)
+        if time.time() >= timeout:
+            raise OSError("Timeout starting {}".format(cls._process_cmd))
```
https://travis-ci.org/github/g1itch/PyBitmessage/jobs/738338187

PeterSurda replied:
That can be fixed later. We don't have a working OSX at the moment anyway.

PeterSurda commented:
> I don't remember any shutdown tests failing in your test environment or current xenial travis setup. `TestAPIShutdown` failed on windows and on my recent fedora31 container.

I can put the timeouts back, maybe some other changes that I made fixed the issues.
 
> Windows failed until this run: https://travis-ci.org/github/g1itch/PyBitmessage/jobs/736655954

Looks like pyelliptic pointers.

PeterSurda commented:
@g1itch I refactored the use of state.appdata and removed the environment variable from .travis.yml. It now passes buildbot's bionic and xenial, as well as xenial on travis-ci.org. Can you try rebasing your changes on top of it again?

g1itch commented:
Wow! It looks like you really fixed the major issue with appdata while I was on a half way retracing it from debug -> helper_startap <-> state (loadConfig).

g1itch commented:
> @g1itch I refactored the use of `state.appdata` and removed the environment variable from `.travis.yml`. It now passes buildbot's bionic and xenial, as well as xenial on travis-ci.org. Can you try rebasing your changes on top of it again?

I do the opposite: I merge it to cleaned out ci branch: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738506280

PeterSurda commented:
> I do the opposite

Whichever you prefer.

g1itch commented on src/tests/test_process.py:
```diff
@@ -56,6 +92,17 @@ def _stop_process(cls, timeout=5):
+        except psutil.NoSuchProcess:
+            return True
+        except psutil.TimeoutExpired:
+            raise OSError("Process is stuck")
```
Maybe `self.fail()` here?

PeterSurda replied:
ok

PeterSurda replied:
`_kill_process` is a class method so you'd have to use `cls.fail()` and I don't think that would work correctly. The documentation doesn't say what happens if there are exceptions in `tearDownClass`, only in `tearDown`. Looking at [unittest source](https://github.com/python/cpython/blob/0f221d09cad46bee38d1b7a7822772df66c53028/Lib/unittest/suite.py), it looks like an exception in `tearDownClass` also adds a new error entry in the totals, just like `tearDown` is supposed to. So I think a more traditional exception like `OSError` is a better choice here. Perhaps an even better choise is to simply delete this `except` and let the `psutil.TimeoutExpired` propagate up.

g1itch commented on src/tests/test_config.py:
```diff
@@ -45,6 +45,7 @@ class TestProcessConfig(TestProcessProto):
     def test_config_defaults(self):
         """Test settings in the generated config"""
         self._stop_process()
+        self._kill_process()
```
This is like `tearDown()` without cleaning the files.

PeterSurda replied:
Yes I think you're correct.

g1itch reviewed (approved):
Some formatting options and two errors on windows I found.

g1itch commented:
> > Windows failed until this run: https://travis-ci.org/github/g1itch/PyBitmessage/jobs/736655954
> 
> Looks like pyelliptic pointers.

That blindsig tests failed because of openssl-1.0.1 dll. If I install PyQt4 like in `winbuild.sh` it will fix the tests, because pyelliptic loads openssl-1.0.1 dll bundled with pyqt in that case.

PeterSurda commented on src/tests/test_process.py:
```diff
@@ -56,6 +92,17 @@ def _stop_process(cls, timeout=5):
+    @classmethod
+    def _kill_process(cls, timeout=5):
+        try:
+            cls.process.send_signal(signal.SIGKILL)
```
ok

PeterSurda replied:
Added an exception handler.

PeterSurda commented on src/tests/test_process.py:
```diff
@@ -69,8 +116,10 @@ def tearDownClass(cls):
                 cls.process.kill()
+                cls.process.wait(5)
+        except psutil.TimeoutExpired:
+            pass
```
ok, it can be split later in case you want to handle them differently

PeterSurda replied:
Done

PeterSurda commented on src/tests/test_process.py:
```diff
@@ -110,7 +191,16 @@ def test_shutdown(self):
+        except psutil.TimeoutExpired:
+            pass
+        except psutil.NoSuchProcess:
+            pass
```
same as above

PeterSurda replied:
done

g1itch commented on src/tests/test_process.py:
```diff
@@ -94,8 +143,40 @@ def _test_threads(self):
+            ",".join(missing_threads), ",".join(extra_threads))
+
+        self.assertGreaterEqual(
+            len(self.process.threads()),
```
`AccessDenied` on osx: https://travis-ci.org/github/g1itch/PyBitmessage/jobs/738524694

PeterSurda replied:
Test is now skipped on OSX.

PeterSurda commented:
@g1itch Thanks for the feedback so far, I'll fix those issues.

g1itch reviewed (approved):
Non-linux issues may be fixed later

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +55,25 @@ class TestProcessProto(unittest.TestCase):
+                    break
+            except OSError:
+                pass
+            time.sleep(1)
```
Not so easy: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738699572

g1itch replied:
Rebased in my favor: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738705081

PeterSurda replied:
I added `os.path.exists`.

g1itch replied:
if not `os.path.exists` it will be full wait until `pybitmessage -t` stops. See:
https://github.com/Bitmessage/PyBitmessage/blob/v0.6/src/bitmessagemain.py#L180
https://github.com/Bitmessage/PyBitmessage/blob/v0.6/src/bitmessagemain.py#L371

g1itch replied:
Hmm https://travis-ci.org/github/g1itch/PyBitmessage/jobs/739258079

g1itch replied:
no fractional part in `st_mtime`: https://travis-ci.org/github/g1itch/PyBitmessage/jobs/739262652

PeterSurda replied:
Ok I'll look throught this.

PeterSurda commented on src/tests/test_process.py:
```diff
@@ -94,8 +144,42 @@ def _test_threads(self):
         #   - Downloader
-        self.assertEqual(
-            len(self.process.threads()), self._threads_count)
+        #   - Uploader
```
Yea makes sense.

g1itch commented on src/tests/test_process.py:
```diff
@@ -94,8 +144,42 @@ def _test_threads(self):
+        self.assertLessEqual(
+            running_threads,
+            self._threads_count_max,
+            msg)
```
```patch
commit 77680a2ef4a9dd16736960e457529a66d7a6856d
Author: Dmitri Bogomolov <4glitch@gmail.com>
Date:   Mon Oct 26 07:57:52 2020 +0200

    Fix test_threads for non-linux environment:
    
     - check found thread_names adequacy, use num_threads()
     - take into account one extra thread on windows

diff --git a/src/tests/test_process.py b/src/tests/test_process.py
index cd114323..63aa2e9c 100644
--- a/src/tests/test_process.py
+++ b/src/tests/test_process.py
@@ -12,8 +12,6 @@ import unittest
 
 import psutil
 
-from sys import platform
-
 from pybitmessage.paths import set_appdata_folder
 
 
@@ -126,49 +124,36 @@ class TestProcessProto(unittest.TestCase):
             cls._cleanup_files()
 
     def _test_threads(self):
-        # only count for now
-        # because of https://github.com/giampaolo/psutil/issues/613
-        # PyBitmessage
-        #   - addressGenerator
-        #   - singleWorker
-        #   - SQL
-        #   - objectProcessor
-        #   - singleCleaner
-        #   - singleAPI
-        #   - Asyncore
-        #   - ReceiveQueue_0
-        #   - ReceiveQueue_1
-        #   - ReceiveQueue_2
-        #   - Announcer
-        #   - InvBroadcaster
-        #   - AddrBroadcaster
-        #   - Downloader
-        #   - Uploader
-
         self.longMessage = True
 
         try:
-            thread_names = \
-                subprocess.check_output(["ps", "-L", "-o", "comm=",
-                                         "--pid",
-                                         str(self.process.pid)]).split()
-        # pylint: disable=broad-except
-        except BaseException:
+            # using ps for posix platforms
+            # because of https://github.com/giampaolo/psutil/issues/613
+            thread_names = subprocess.check_output([
+                "ps", "-L", "-o", "comm=", "--pid",
+                str(self.process.pid)
+            ]).split()
+        except:  # pylint: disable=broad-except
             thread_names = []
 
-        extra_threads = []
-        missing_threads = []
-        for thread_name in thread_names:
-            if thread_name not in self._threads_names:
-                extra_threads.append(thread_name)
-        for thread_name in self._threads_names:
-            if thread_name not in thread_names:
-                missing_threads.append(thread_name)
-
-        msg = "Missing threads: {}, Extra threads: {}".format(
-            ",".join(missing_threads), ",".join(extra_threads))
-
-        running_threads = len(self.process.threads())
+        running_threads = len(thread_names)
+        if 0 < running_threads < 30:  # adequacy check
+            extra_threads = []
+            missing_threads = []
+            for thread_name in thread_names:
+                if thread_name not in self._threads_names:
+                    extra_threads.append(thread_name)
+            for thread_name in self._threads_names:
+                if thread_name not in thread_names:
+                    missing_threads.append(thread_name)
+
+            msg = "Missing threads: {}, Extra threads: {}".format(
+                ",".join(missing_threads), ",".join(extra_threads))
+        else:
+            running_threads = self.process.num_threads()
+            if sys.platform.startswith('win'):
+                running_threads -= 1  # one extra thread on Windows!
+            msg = "Unexpected running thread count"
 
         self.assertGreaterEqual(
             running_threads,
@@ -220,7 +205,6 @@ class TestProcess(TestProcessProto):
                 'Failed to read file %s' % pfile
             )
 
-    @unittest.skipIf(platform.startswith('darwin'), "Doesn't work on OSX")
     def test_threads(self):
         """Testing PyBitmessage threads"""
         self._test_threads()

```

g1itch replied:
Almost clear: https://travis-ci.org/github/g1itch/PyBitmessage/builds/738955127

PeterSurda replied:
I added this now.

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +54,27 @@ class TestProcessProto(unittest.TestCase):
+                    break
+            time.sleep(1)
+        if time.time() >= timeout:
+            raise psutil.TimeoutExpired("Timeout starting {}".format(cls._process_cmd))
```
87 char line

PeterSurda replied:
I see.

g1itch reviewed (approved):
TImeout expired every time on osx: https://travis-ci.org/github/g1itch/PyBitmessage/jobs/739252746

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +54,27 @@ class TestProcessProto(unittest.TestCase):
+        starttime = time.time()
         subprocess.call(cls._process_cmd)  # nosec
-        time.sleep(5)
+        timeout = starttime + 30.0
```
milliseconds?

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +54,27 @@ class TestProcessProto(unittest.TestCase):
+        cls.home = set_appdata_folder(tempfile.gettempdir())
+        cls._cleanup_files()
         put_signal_file(cls.home, 'unittest.lock')
+        starttime = time.time()
```
maybe `int(time.time())`?

PeterSurda replied:
Ok makes sense.

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +54,27 @@ class TestProcessProto(unittest.TestCase):
+                    if starttime <= pstat.st_mtime and pstat.st_size > 0:
+                        break
+                except OSError:
+                    break
```
As I can see `OSError` raised when `os.path.exists(os.path.join(cls.home, 'singleton.lock')) is False`.

PeterSurda replied:
Oh, now I get it.

g1itch commented on src/tests/test_process.py:
```diff
@@ -32,10 +54,27 @@ class TestProcessProto(unittest.TestCase):
+            raise psutil.TimeoutExpired("Timeout starting {}".format(cls._process_cmd))
+        # wait a bit for the program to fully start
+        # 10 sec should be enough
+        time.sleep(10)
```
Why sleep 10 seconds if singleton.lock is already checked? What may happen?

g1itch replied:
OK, I see. It is to start all threads and not fail test_threads and test_shutdown.

g1itch replied:
But I hope that 5 sec will be enough for any environment (on travis it works without delay). This delay is used before starting tests in any test case subclassed from `TestProcessProto` on each pass.

g1itch commented:
No handlers could be found for logger "default": https://travis-ci.org/github/Bitmessage/PyBitmessage/builds/739215140#L1084

g1itch commented on src/tests/test_logger.py:
```diff
@@ -37,33 +39,44 @@ class TestLogger(unittest.TestCase):
+                         format(pass_))
 
-        self.assertRegexpMatches(open(log_file).read(), pattern)
+        logger2 = logging.getLogger('root')
```
I don't understand the purpose, it's not used in code:
```shell
$ grep -r 'getLogger' src | grep root
src/tests/test_logger.py:        logger2 = logging.getLogger('root')
```

PeterSurda replied:
I think it may be leftover from debugging, perhaps it isn't necessary anymore.

g1itch commented on src/tests/test_logger.py:
```diff
@@ -37,33 +39,44 @@ class TestLogger(unittest.TestCase):
+                                 ", pass {}".format(pass_))
+
+        if pass_ == 2:
+            os.remove(log_config)
```
Why only on second pass?

PeterSurda replied:
I don't understand the logic flow of the passes, this may be a leftover from debugging and perhaps isn't necesasry anymore.

g1itch replied:
This not fixes the test. I'll describe how it works, how it worked and how I wanted it to work a bit later.

PeterSurda replied:
I'll look at it.

g1itch replied:
> I don't understand the logic flow of the passes, this may be a leftover from debugging and perhaps isn't necesasry anymore.

A proper test should not rely on other tests side effects.

What I wanted to demonstrate in this particular test:
1. when you import `logger` from `debug` it sets up the logging based on `logging.dat` config
1. `logger = logging.getLogger('root')` set after the import will be the same as in `debug`
1. if you call `resetLogging()` the configuration will be reapplied

All that works only after the first import of `debug` because it calls `helper_startup.loadConfig()`.

How it worked before:
1. precondition: there is no other test importing `debug` or other module inducing config setup or appdata change
1. first pass detected by `debug.log` in appdata remaining from core tests
1. on the second pass you need `resetLogging()`

You can reveal it by replacing all `tempfile.gettempdir()` by `tempfile.mkdtemp()` as was suggested before.

g1itch replied:
I'm not sure that I can precisely reconstruct what's happening now. It places `logging.dat` which is used by subsequent runs of pybitmessage process and can be detected on second pass.

g1itch replied:
Maybe subclass this test case from `TestProcessProto` and check only the presence of pattern and maybe severity and initial message? Then we can replace `logging.getLogger('default')` by conventional `logging.getLogger(__name__)`, compose proper `logging.dat` and check presence of all submodule names which write logs.

PeterSurda replied:
> first pass detected by debug.log in appdata remaining from core tests

Only if it uses the same config directory. But the tests change the location of the config directory. I think this breaks the logic. You can see now that I reverted to your original logic in `test_logger.py`, and it fails on travis-ci.org. I don't want to merge until this is resolved.

g1itch replied:
I wrote before that my original logic was wrong. Standalone tests should not import any module with side effects.
I debug my process-based variant, it succedes only on linux: https://travis-ci.org/github/g1itch/PyBitmessage/builds/740134631

PeterSurda commented:
I'll work on the requested changes.

g1itch commented:
I'm waiting to check the new approach: https://travis-ci.org/github/g1itch/PyBitmessage/builds/739605155

g1itch commented:
The cleanup of the core tests is [here](https://github.com/Bitmessage/PyBitmessage/blob/v0.6/src/tests/core.py#L51).

g1itch commented:
xenial, bionic and fedora31: OK, small hangup on focal, waiting for windows

https://travis-ci.org/github/g1itch/PyBitmessage/builds/739657803 - process based `TestLogger` and cleanup tweaks.

PeterSurda commented:
I think all your feedback has been applied, and travis-ci passes. I think the rest of the issue can be resolved later, and this is ok to merge now. Let me know if you agree.

g1itch commented:
> I think all your feedback has been applied, and travis-ci passes. I think the rest of the issue can be resolved later, and this is ok to merge now. Let me know if you agree.

Now I don't see any help from `get_active_config_folder()` and related code.

PeterSurda commented:
> Now I don't see any help from get_active_config_folder() and related code.

It works on xenial on buildbot now.

PeterSurda commented:
So what now?

g1itch commented:
> > Now I don't see any help from get_active_config_folder() and related code.
> 
> It works on xenial on buildbot now.

But it you remove initial `src/bitmessagemain.py -t` it may stop working.

PeterSurda commented:
Well you can fix it later.

PeterSurda commented:
So apart from the logger do you have any other complaints?

g1itch commented:
> So apart from the logger do you have any other complaints?

While this version `TestLogger` is here any newly written test case may fail unexpectedly.

`get_active_config_folder()` code is doing nothing.

Also I have one more assertion which should not fail. I'll prepare another draft PR so you can test it on your buildbot.

PeterSurda commented:
> get_active_config_folder() code is doing nothing.

It fixes the `test_config.py` errors. The logger tests can be fixed in a separate PR.

g1itch commented:
> > get_active_config_folder() code is doing nothing.
> 
> It fixes the `test_config.py` errors. The logger tests can be fixed in a separate PR.

What fixes `test_config` is proper cleanup and guaranteed kill in `tearDownClass` - that's what I'm going to demonstrate in the new rebased branch.

PeterSurda commented:
Ok then I close this and you can make a PR with your changes.
